crates/pop-cli/src/commands/call/parachain.rs

```diff
@@ -0,0 +1,813 @@
+			Ok(tx) => tx,
+			Err(e) => {
+				return Err(anyhow!("Error: {}", e));
+			},
+		};
+		cli.info(format!("Encoded call data: {}", encode_call_data(client, &tx)?))?;
+		Ok(tx)
+	}
+
+	// Sign and submit an extrinsic.
+	async fn submit_extrinsic(
+		&mut self,
+		client: &OnlineClient<SubstrateConfig>,
+		tx: DynamicPayload,
+		cli: &mut impl Cli,
+	) -> Result<()> {
+		if !self.skip_confirm &&
+			!cli.confirm("Do you want to submit the extrinsic?")
+				.initial_value(true)
+				.interact()?
+		{
+			display_message(
+				&format!(
+					"Extrinsic {} was not submitted. Operation canceled by the user.",
+					self.extrinsic.name
+				),
+				false,
+				cli,
+			)?;
+			return Ok(());
+		}
+		let spinner = cliclack::spinner();
+		spinner.start("Signing and submitting the extrinsic, please wait...");
+		let result = sign_and_submit_extrinsic(client.clone(), tx, &self.suri)
+			.await
+			.map_err(|err| anyhow!("{}", format!("{err:?}")))?;
+
+		spinner.stop(format!("Extrinsic submitted with hash: {:?}", result));
+		Ok(())
+	}
+	fn display(&self, chain: &Chain) -> String {
+		let mut full_message = "pop call parachain".to_string();
+		full_message.push_str(&format!(" --pallet {}", self.pallet));
+		full_message.push_str(&format!(" --extrinsic {}", self.extrinsic));
+		if !self.args.is_empty() {
+			let args: Vec<_> = self.args.iter().map(|a| format!("\"{a}\"")).collect();
+			full_message.push_str(&format!(" --args {}", args.join(" ")));
+		}
+		full_message.push_str(&format!(" --url {} --suri {}", chain.url, self.suri));
+		full_message
+	}
+}
+
+fn display_message(message: &str, success: bool, cli: &mut impl Cli) -> Result<()> {
+	if success {
+		cli.outro(message)?;
+	} else {
+		cli.outro_cancel(message)?;
+	}
+	Ok(())
+}
+
+// Prompts the user for some predefined actions.
+async fn prompt_predefined_actions(
+	pallets: &[Pallet],
+	cli: &mut impl Cli,
+) -> Result<Option<Action>> {
+	let mut predefined_action = cli.select("What would you like to do?");
+	for action in supported_actions(pallets).await {
+		predefined_action = predefined_action.item(
+			Some(action.clone()),
+			action.description(),
+			action.pallet_name(),
+		);
+	}
+	predefined_action = predefined_action.item(None, "All", "Explore all pallets and extrinsics");
+	Ok(predefined_action.interact()?)
+}
+
+// Prompts the user for the value of a parameter.
+fn prompt_for_param(
+	client: &OnlineClient<SubstrateConfig>,
+	cli: &mut impl Cli,
+	param: &Param,
+) -> Result<String> {
+	if param.is_optional {
+		if !cli
+			.confirm(format!(
+				"Do you want to provide a value for the optional parameter: {}?",
+				param.name
+			))
+			.interact()?
+		{
+			return Ok("None()".to_string());
+		}
+		let value = get_param_value(client, cli, param)?;
+		Ok(format!("Some({})", value))
+	} else {
+		get_param_value(client, cli, param)
+	}
+}
+
+// Resolves the value of a parameter based on its type.
+fn get_param_value(
+	client: &OnlineClient<SubstrateConfig>,
+	cli: &mut impl Cli,
```

Comment:
Not sure these are necessary

Reply:
This would allow to simplify the code and separate the tests much better.

Reply:
What are you suggesting here? To remove the auxiliary methods? 
I could include all the logic directly in the main method. However, splitting it into auxiliary methods feels cleaner and improves readability no? Open for to refactor it if is clearer the other way.

Reply:
`client` and `cli` parameters are not used if I don't mistake

Reply:
Understood now, I thought you were referring to the auxiliary functions. You’re right, the client isn’t needed, only the cli parameter is. Fixed in https://github.com/r0gue-io/pop-cli/pull/316/commits/28743bd44c1f222880222c2a0aa29363275f1f29

Reply:
You sure the `cli` is needed? I thought that one was also not needed

Reply:
Yes, all the subfunctions need to either prompt the user or recursively call the main function with the cli parameter.